### PR TITLE
first commit

### PR DESCRIPTION
**Pull request by monicaafilip**

### CODE REVIEW

File: Game/MyProject/Game/Helper/HelperClass.cpp [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    {
+        BuildQuad(meshInstance);
+        builtQuad=true;
+        
+        IvImage *image=IvImage::CreateFromFile(textureName);
+        if(image)
+        {
+            quadTexture=IvRenderer::mRenderer->GetResourceManager()
+            ->CreateTexture((image->GetBytesPerPixel()==4) ? kRGBA32TexFmt : kRGB24TexFmt,image
+                            ->GetWidth(), image->GetHeight(), image->GetPixels(), kImmutableUsage);
```
Comment (mircea-catana) [inline Game/MyProject/Game/Helper/HelperClass.cpp]: writing code like this is very unreadable and prone to human error

File: Game/MyProject/Game/Helper/HelperClass.cpp [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            delete image;
+            image = 0;
+        }
+    }
+    IvShaderProgram* shader=meshInstance->getShader();
+    IvUniform* unif=shader->GetUniform("myTexture");
```
Comment (mircea-catana) [inline Game/MyProject/Game/Helper/HelperClass.cpp]: if you want to use this with a different shader you might not find "myTexture" as a uniform

File: Game/MyProject/Game/Helper/HelperManager.cpp [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+#include"HelperManager.h"
+//-------------------------------------------------------------------------------
+//  PUBLIC METHODS
+//-------------------------------------------------------------------------------
+
+//-------------------------------------------------------------------------------
+// @ HelperManager::CreateBoxMesh()
+//-------------------------------------------------------------------------------
+std::shared_ptr<Mesh>
+HelperManager::CreateBoxMesh()
```
Comment (mircea-catana) [inline Game/MyProject/Game/Helper/HelperManager.cpp]: this is not how you used indexed drawing. i'm counting 24 vertices and redundant indexing that i don't even think works. doing indexed drawing like this is actually slower than non indexed. a cube should have 8 vertices and a proper index buffer to generate the triangles from those 8 vertices.

File: Game/MyProject/Game/Helper/HelperManager.cpp [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+    vertex1.position=points[0];
+    vertex2.position=points[11];
+    vertex3.position=points[5];
+    
+    vertices.push_back(vertex1);
+    vertices.push_back(vertex2);
+    vertices.push_back(vertex3);
+    
+    std::vector<IvTNPVertex> vertices2;
+    vertices2=RefineTriangles(vertices, 4);
```
Comment (mircea-catana) [inline Game/MyProject/Game/Helper/HelperManager.cpp]: refinement level should be configurable

File: Game/MyProject/Game/Helper/HelperManager.cpp [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+    sphereMesh->setVertexBuffer(vertices2, format);
+    
+    std::vector<unsigned int> indices;
+    for(int i=0;i<vertices2.size();i++)
+        indices.push_back(i);
```
Comment (mircea-catana) [inline Game/MyProject/Game/Helper/HelperManager.cpp]: again, very bad use of index buffers, just like the box one.
this whole method looks weird with ~150lines of specific vertex pushing.
you should be able to generate the needed vertex positions in a loop using polar coordinates.

File: Game/MyProject/Game/Helper/HelperManager.cpp [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+HelperManager::GetMiddlePoint(IvVector3 p1, IvVector3 p2)
+{
+    IvVector3 mid=IvVector3(    (p1.x+p2.x)/2.0,
+                                (p1.y+p2.y)/2.0,
+                                (p1.z+p2.z)/2.0);
+    
+    return mid;
+}
+//-------------------------------------------------------------------------------
+// @ HelperManager::GetMiddlePoint()
```
Comment (mircea-catana) [inline Game/MyProject/Game/Helper/HelperManager.cpp]: fix doc pls

File: Game/MyProject/Game/Helper/HelperManager.cpp [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+//-------------------------------------------------------------------------------
+std::vector<IvTNPVertex>
+HelperManager::RefineTriangles(std::vector<IvTNPVertex> vertices,int recursionLevel)
```
Comment (mircea-catana) [inline Game/MyProject/Game/Helper/HelperManager.cpp]: I'm not so sure about this approach. You could just generate all the vertices upfront in the CreateSphereMesh method and avoid this nested-loop recursion.. but if you fix the other things and it works ok you can keep it..

File: Game/MyProject/Game/Test/TestHelper/TestHelper.cpp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    createdQuads.push_back(false);
+    createdQuads.push_back(false);
+   
+    
+}
+
+void
+TestHelper::Draw()
+{
+//    //test boxes
```
Comment (mircea-catana) [inline Game/MyProject/Game/Test/TestHelper/TestHelper.cpp]: don't commit blocks of commented code.
if you want to keep it around for later testing move it to a function and just don't call it right now..

File: Game/MyProject/IvGraphics/IvVertexFormats.h [hunk 9]
```diff
@@ -101,13 +101,14 @@ T* VertexDescription::NextVertexAttribute(T * attributePtr)
...
-    kTCPFormat,     // texture coord, color, position
-    kTNPFormat,      // texture coord, normal, position
-
+    kPFormat,               // position only
+    kCPFormat,              // color, position
+    kNPFormat,              // normal, position
+    kCNPFormat,             // color, normal, position
+    kTCPFormat,             // texture coord, color, position
+    kTNPFormat,             // texture coord, normal, position
+    kVertexDescription,     // anything we want
```
Comment (mircea-catana) [inline Game/MyProject/IvGraphics/IvVertexFormats.h:110]: i want a dragon 🐉

**Review by mircea-catana — CHANGES_REQUESTED** [review]
there's some cleaning up to do here..

File: Game/MyProject/Game/Helper/HelperManager.h [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    std::shared_ptr<MeshInstance> CreateMeshInstance(const std::shared_ptr<Mesh>&,const std::vector<std::string>& uniforms,const char* shaderName="TextureShader");
+    
+    std::shared_ptr<SceneNode> BuildBox(const std::shared_ptr<MeshInstance>&,bool wireframeValue=false);
+    std::shared_ptr<SceneNode> BuildSphere(const std::shared_ptr<MeshInstance>&,bool wireframeValue=false);
+    std::shared_ptr<SceneNode> BuildQuad(const std::shared_ptr<MeshInstance>&,IvVector3 axis=IvVector3::zAxis, bool wireframeValue=false);
+    std::shared_ptr<SceneNode> BuildTexturedQuad(const std::shared_ptr<MeshInstance>&,const char* textureName,const char* shaderName="TextureShader",IvVector3 axis=IvVector3::zAxis, bool wireframeValue=false);
+    
+    IvVector3 GetMiddlePoint(IvVector3 p1,IvVector3 p2);
+    void RefineTriangles(std::vector<IvTNPVertex>& vertices,std::vector<unsigned int>& indices,int recursionLevel);
+    float CalculateLength(IvVector3 point);
```
Comment (krahen94) [inline Game/MyProject/Game/Helper/HelperManager.h]: const IvVector3&

File: Game/MyProject/Game/Helper/MeshManager.cpp [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+{
+    std::map<std::string,bool>::iterator iterator;
+    iterator=createdMeshes.find(name);
+    
+    if(iterator==createdMeshes.end())
+    {
+        std::shared_ptr<Mesh> mesh;
+
+        if(name=="box")
+
```
Comment (krahen94) [inline Game/MyProject/Game/Helper/MeshManager.cpp]: Better formatting here please

File: Game/MyProject/Game/Helper/MeshManager.h [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+//-------------------------------------------------------------------------------
+//-- Classes --------------------------------------------------------------------
+//-------------------------------------------------------------------------------
+class MeshManager
+{
+public:
+    MeshManager(){}
+    ~MeshManager(){}
+    
+    std::shared_ptr<Mesh> GetMesh(std::string name);
```
Comment (krahen94) [inline Game/MyProject/Game/Helper/MeshManager.h]: const std::string&

File: Game/MyProject/Game/Helper/MyHelperSceneNode.cpp [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+{
+    myRenderPacket=renderPacket;
+}
+//-------------------------------------------------------------------------------
+// @ MySceneNodeHelper::collectRenderPacket()
+//-------------------------------------------------------------------------------
+void
+MyHelperSceneNode::collectRenderingPackets(CameraSceneNode * camera, std::vector<RenderPacket> & renderPackets)
+{
+    if (SceneNode::_rendarable) {
```
Comment (krahen94) [inline Game/MyProject/Game/Helper/MyHelperSceneNode.cpp]: _rendarable -> isRenderable or something

File: Game/MyProject/Game/Helper/MyHelperSceneNode.h [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+//-------------------------------------------------------------------------------
+//-- Classes --------------------------------------------------------------------
+//-------------------------------------------------------------------------------
+
+class MyHelperSceneNode:public SceneNode
+{
+public:
+	
+    MyHelperSceneNode(const std::string& name,RenderPacket renderPacket);
+    ~MyHelperSceneNode(){}
```
Comment (krahen94) [inline Game/MyProject/Game/Helper/MyHelperSceneNode.h]: When doing destructors in an inheritance chain, you either mark it virtual, or final

File: Game/MyProject/Game/Helper/MyHelperSceneNode.h [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    
+    void collectRenderingPackets(CameraSceneNode*,std::vector<RenderPacket>&);
+
+private:
+
+    RenderPacket       myRenderPacket;
```
Comment (krahen94) [inline Game/MyProject/Game/Helper/MyHelperSceneNode.h]: Better naming conventions and formatting please.

m_renderPacket or just renderPacket works.

**Review by krahen94 — CHANGES_REQUESTED** [review]
It looks mostly good, but please enforce proper formatting and stick to one style of variable naming.

File: Game/MyProject/Game/Helper/HelperManager.cpp [hunk 16]
```diff
@@ -0,0 +1,553 @@
...
+        indices.push_back(6);
+        indices.push_back(7);
+        
+        indices.push_back(9);
+        indices.push_back(8);
+        indices.push_back(1);
+        
+        indices.push_back(10);
+        indices.push_back(11);
+        indices.push_back(5);
```
Comment (mircea-catana) [inline Game/MyProject/Game/Helper/HelperManager.cpp]: i'm pretty sure you can find a pattern and put all of this in a loop.
this code smells (https://en.wikipedia.org/wiki/Code_smell)

File: Game/MyProject/Game/Helper/HelperManager.cpp [hunk 17]
```diff
@@ -0,0 +1,553 @@
...
+        
+        
+        RefineTriangles(vertices,indices,4);
```
Comment (mircea-catana) [inline Game/MyProject/Game/Helper/HelperManager.cpp]: refinement level still not exposed to api

File: Game/MyProject/Game/Helper/HelperManager.cpp [hunk 18]
```diff
@@ -0,0 +1,553 @@
...
+    }
+    //---------------------------------------------------------------------------
+    // @HelperManager::BuildSphere()
+    //---------------------------------------------------------------------------
+    std::shared_ptr<SceneNode>
+    BuildSphere(const std::shared_ptr<MeshInstance>&meshInstance,bool wireframeValue)
+    {
+        RenderPacket myRenderPacket;
+        myRenderPacket._use_blend = true;
+        myRenderPacket._use_depth = true;
```
Comment (mircea-catana) [inline Game/MyProject/Game/Helper/HelperManager.cpp]: these are generic primitive builder methods.. are you sure blend and depth testing will be enabled for all their uses?

File: Game/MyProject/Game/Helper/HelperManager.cpp [hunk 19]
```diff
@@ -0,0 +1,553 @@
...
+        else if (axis == IvVector3::zAxis)
+            rotation += IvVector3(0, 0, 1.5);
+        node->setLocalTransform(position, rotation, scale);
+        
+        IvImage* image = IvImage::CreateFromFile(textureName);
+        IvTexture* quadTexture;
+        
+        if (image)
+        {
+            quadTexture = IvRenderer::mRenderer->GetResourceManager()->CreateTexture((image->GetBytesPerPixel() == 4) ? kRGBA32TexFmt : kRGB24TexFmt, image->GetWidth(), image->GetHeight(), image->GetPixels(), kImmutableUsage);
```
Comment (mircea-catana) [inline Game/MyProject/Game/Helper/HelperManager.cpp]: too much stuff going on in this line of code, break stuff up so it's readable. see the code smell url above

File: Game/MyProject/Game/Helper/HelperManager.cpp [hunk 20]
```diff
@@ -0,0 +1,553 @@
...
+            delete image;
+            image = 0;
+        }
+        
+        IvUniform* unif = meshInstance->getShaderUniforms()[0];
```
Comment (mircea-catana) [inline Game/MyProject/Game/Helper/HelperManager.cpp:439]: are you sure uniforms[0] is always going to be the texture? looks shady to me

File: Game/MyProject/Game/Helper/MeshManager.hpp [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+//-- Dependencies ---------------------------------------------------------------
+//-------------------------------------------------------------------------------
+#include "Mesh.hpp"
+#include "HelperManager.hpp"
+#include <stdio.h>
+#include <map>
+//-------------------------------------------------------------------------------
+//-- Classes --------------------------------------------------------------------
+//-------------------------------------------------------------------------------
+class MeshManager
```
Comment (mircea-catana) [inline Game/MyProject/Game/Helper/MeshManager.hpp:15]: what's the purpose of this class?

**Review by raxvan — APPROVED** [review]

File: Game/MyProject/Game/Helper/MeshManager.hpp (hunk 21 above)
Comment (monicaafilip) [reply Game/MyProject/Game/Helper/MeshManager.hpp:15]: I need only a mesh for a sphere,box,quad and many instances. So this class keeps track of it. I create the mesh only once and use it.

Comment (mircea-catana) [reply Game/MyProject/Game/Helper/MeshManager.hpp:15]: sure, or you could just keep a static reference in the get method and skip the whole get by name and additional class. anyway, let's not waste time redoing this, it's fine

**Pull request merged**